<a href="https://colab.research.google.com/github/elenasoria3/tareas_master_PLN/blob/main/Evaluacion_traducciones_ESL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Evaluación automática de traducciones


In [ ]:
original = "Trash Can Nachos is what the new Netflix series Ginny & Georgia most reminds me of. It’s an hourlong drama about a woman named Georgia (Brianne Howey) with a checkered past and two kids, who moves to an affluent suburban community in Massachusetts for a fresh start. Her daughter, Ginny (Antonia Gentry), was born when Georgia was only 15, and the Gilmore Girls mom-and-daughter-are-more-like-sisters impression is palpable. It also seems purposeful, given the double-G hit of the title and the quirky town setting. The Gilmore Girls comparison is obvious and Ginny & Georgia invites it readily, but that reference point is merely one layer of what Ginny & Georgia is doing. To call it a Gilmore knockoff would be like trying to describe Trash Can Nachos by saying, “Oh yeah, they’re like most nachos you’d make for yourself on a typical day.” Because Ginny & Georgia has so much more going on than just the tense dynamics between a mother and daughter who are unusually close in age. As the show starts digging into Georgia’s past, which includes a smorgasbord of illegal activity, Georgia also starts covering her tracks in the present, and the show becomes a crime drama. When Ginny starts making friends at her new school, the show shifts into light teen drama mode, but when they party pretty hard and Ginny engages in some frightening self-harm, the teen drama becomes much darker. Georgia starts working for the town’s mayor (Scott Porter) and in those portions the series shifts more into a local political dramedy. Throughout it all, the series is heavily influenced by who among the several love interests Ginny and Georgia are most drawn to, and in those sections it’s comedy, with a strong dash of brooding romance."
referencia = "Lo que más me recuerda a la nueva serie de Netflix Ginny & Georgia es una receta de Trash Can Nachos. Este drama de una hora trata sobre una mujer, llamada Georgia, Brianne Howey, con un pasado accidentado y dos hijos, que se muda a una acomodada zona de la periferia de Massachussets para comenzar de nuevo. Su hija Ginny, Antonia Gentry, nació cuando Georgia solo tenía 15 años, por lo que el guiño a Gilmore Girls de esta relación entre madre e hija que parecen más hermanas es patente. Es más, parece hecha a propósito si se tiene en cuenta la doble «G» del título y su ambientación en un pueblo peculiar. La comparación con Gilmore Girls es obvia y Ginny & Georgia lo hace ver claramente, pero esta referencia solo constituye una capa de lo que es Ginny & Georgia. Clasificarla como una imitación de Gilmore sería como intentar describir la receta de los Trash Can Nachos diciendo: «Sí, claro, son como la mayoría de nachos que te prepararías en un día cualquiera». Ginny & Georgia va mucho más allá de una tensa dinámica entre madre e hija que se acercan más de lo normal en su edad. Como la serie comienza ahondando en el pasado de Georgia, compuesto por una surtida mezcla de actividades ilegales, Georgia va cubriendo sus huellas en el presente, haciendo que la serie se convierta en un drama criminal. Cuando Ginny empieza a hacer amigos en su nuevo instituto, la serie pasa a ser un liviano drama adolescente, pero cuando no paran de irse de fiesta y Ginny comienza alarmantemente a autolesionarse, el drama adolescente se torna mucho más oscuro. Georgia empieza a trabajar para el alcalde del pueblo (Scott Porter) y en estas escenas la serie parece más una tragicomedia política local. Entre todo esto, la serie está altamente influenciada por los variados intereses amorosos que van captando más la atención de Ginny y Georgia, partes en las cuales se convierte en comedia, con un gran toque de romance cautivador."

In [ ]:
frasesOriginal = [x.strip() for x in original.split('.')[:-1]]
frasesOriginal

['Trash Can Nachos is what the new Netflix series Ginny & Georgia most reminds me of',
 'It’s an hourlong drama about a woman named Georgia (Brianne Howey) with a checkered past and two kids, who moves to an affluent suburban community in Massachusetts for a fresh start',
 'Her daughter, Ginny (Antonia Gentry), was born when Georgia was only 15, and the Gilmore Girls mom-and-daughter-are-more-like-sisters impression is palpable',
 'It also seems purposeful, given the double-G hit of the title and the quirky town setting',
 'The Gilmore Girls comparison is obvious and Ginny & Georgia invites it readily, but that reference point is merely one layer of what Ginny & Georgia is doing',
 'To call it a Gilmore knockoff would be like trying to describe Trash Can Nachos by saying, “Oh yeah, they’re like most nachos you’d make for yourself on a typical day',
 '” Because Ginny & Georgia has so much more going on than just the tense dynamics between a mother and daughter who are unusually close in

In [ ]:
frasesReferencia = [x.strip() for x in referencia.split('.')[:-1]]
frasesReferencia

['Lo que más me recuerda a la nueva serie de Netflix Ginny & Georgia es una receta de Trash Can Nachos',
 'Este drama de una hora trata sobre una mujer, llamada Georgia, Brianne Howey, con un pasado accidentado y dos hijos, que se muda a una acomodada zona de la periferia de Massachussets para comenzar de nuevo',
 'Su hija Ginny, Antonia Gentry, nació cuando Georgia solo tenía 15 años, por lo que el guiño a Gilmore Girls de esta relación entre madre e hija que parecen más hermanas es patente',
 'Es más, parece hecha a propósito si se tiene en cuenta la doble «G» del título y su ambientación en un pueblo peculiar',
 'La comparación con Gilmore Girls es obvia y Ginny & Georgia lo hace ver claramente, pero esta referencia solo constituye una capa de lo que es Ginny & Georgia',
 'Clasificarla como una imitación de Gilmore sería como intentar describir la receta de los Trash Can Nachos diciendo: «Sí, claro, son como la mayoría de nachos que te prepararías en un día cualquiera»',
 'Ginny & G

In [ ]:
traduccion_deepl = "A lo que más me recuerda la nueva serie de Netflix Ginny & Georgia es a Trash Can Nachos. Es un drama de una hora de duración sobre una mujer llamada Georgia (Brianne Howey) con un pasado accidentado y dos hijos, que se muda a una comunidad suburbana acomodada en Massachusetts para empezar de nuevo. Su hija, Ginny (Antonia Gentry), nació cuando Georgia tenía sólo 15 años, y la impresión de que la madre y la hija de Las chicas Gilmore son como hermanas es palpable. También parece intencionada, dado el golpe de doble G del título y la peculiar ambientación de la ciudad. La comparación con Gilmore Girls es obvia y Ginny & Georgia la invita a ello, pero ese punto de referencia no es más que una capa de lo que Ginny & Georgia está haciendo. Llamarlo una imitación de Gilmore sería como intentar describir Trash Can Nachos diciendo: 'Ah, sí, son como la mayoría de los nachos que te harías en un día normal'. Porque Ginny & Georgia va mucho más allá de la tensa dinámica entre una madre y una hija de edad inusualmente cercana. A medida que la serie empieza a indagar en el pasado de Georgia, que incluye un sinfín de actividades ilegales, Georgia también empieza a cubrir sus huellas en el presente, y la serie se convierte en un drama criminal. Cuando Ginny empieza a hacer amigos en su nueva escuela, la serie se convierte en un drama adolescente ligero, pero cuando salen de fiesta y Ginny se autolesiona, el drama adolescente se vuelve mucho más oscuro. Georgia empieza a trabajar para el alcalde de la ciudad (Scott Porter) y en esos momentos la serie se convierte en un drama político local. A lo largo de todo el episodio, la serie se ve fuertemente influenciada por los intereses amorosos que más atraen a Ginny y Georgia, y en esos tramos es comedia, con una fuerte pizca de melancólico romance."
traduccion_yandex = "Nachos de bote de basura es lo que más me recuerda la nueva serie de Netflix Ginny & Georgia. Es un drama de una hora de duración sobre una mujer llamada Georgia (Brianne Howey) con un pasado accidentado y dos hijos, que se muda a una comunidad suburbana acomodada en Massachusetts para comenzar de nuevo. Su hija, Ginny (Antonia Gentry), nació cuando Georgia tenía solo 15 años, y la impresión de que la madre y la hija de Gilmore Girls son más como hermanas es palpable. También parece útil, dado el éxito de doble G del título y el peculiar entorno de la ciudad. La comparación de Gilmore Girls es obvia y Ginny & Georgia la invita fácilmente, pero ese punto de referencia es simplemente una capa de lo que Ginny & Georgia está haciendo. Llamarlo una imitación de Gilmore sería como tratar de describir los Nachos de bote de basura diciendo: 'Oh, sí, son como la mayoría de los nachos que harías para ti en un día típico.' Porque Ginny & Georgia tiene mucho más que hacer que solo la dinámica tensa entre una madre y una hija que son inusualmente cercanas en edad. A medida que el programa comienza a indagar en el pasado de Georgia, que incluye una mezcla heterogénea de actividades ilegales, Georgia también comienza a cubrir sus huellas en el presente, y el programa se convierte en un drama criminal. Cuando Ginny comienza a hacer amigos en su nueva escuela, el programa cambia al modo de drama adolescente ligero, pero cuando se divierten bastante y Ginny se autolesiona, el drama adolescente se vuelve mucho más oscuro. Georgia comienza a trabajar para el alcalde de la ciudad (Scott Porter) y en esas partes la serie cambia más hacia una comedia dramática política local. A lo largo de todo, la serie está fuertemente influenciada por quién de los varios intereses amorosos a los que Ginny y Georgia se sienten más atraídos, y en esas secciones es comedia, con una fuerte pizca de romance inquietante."

In [ ]:
frasesdeepl = [x.strip() for x in traduccion_deepl.split('.')[:-1]]
frasesdeepl

['A lo que más me recuerda la nueva serie de Netflix Ginny & Georgia es a Trash Can Nachos',
 'Es un drama de una hora de duración sobre una mujer llamada Georgia (Brianne Howey) con un pasado accidentado y dos hijos, que se muda a una comunidad suburbana acomodada en Massachusetts para empezar de nuevo',
 'Su hija, Ginny (Antonia Gentry), nació cuando Georgia tenía sólo 15 años, y la impresión de que la madre y la hija de Las chicas Gilmore son como hermanas es palpable',
 'También parece intencionada, dado el golpe de doble G del título y la peculiar ambientación de la ciudad',
 'La comparación con Gilmore Girls es obvia y Ginny & Georgia la invita a ello, pero ese punto de referencia no es más que una capa de lo que Ginny & Georgia está haciendo',
 "Llamarlo una imitación de Gilmore sería como intentar describir Trash Can Nachos diciendo: 'Ah, sí, son como la mayoría de los nachos que te harías en un día normal'",
 'Porque Ginny & Georgia va mucho más allá de la tensa dinámica entre

In [ ]:
frasesyandex = [x.strip() for x in traduccion_yandex.split('.')[:-1]]
frasesyandex

['Nachos de bote de basura es lo que más me recuerda la nueva serie de Netflix Ginny & Georgia',
 'Es un drama de una hora de duración sobre una mujer llamada Georgia (Brianne Howey) con un pasado accidentado y dos hijos, que se muda a una comunidad suburbana acomodada en Massachusetts para comenzar de nuevo',
 'Su hija, Ginny (Antonia Gentry), nació cuando Georgia tenía solo 15 años, y la impresión de que la madre y la hija de Gilmore Girls son más como hermanas es palpable',
 'También parece útil, dado el éxito de doble G del título y el peculiar entorno de la ciudad',
 'La comparación de Gilmore Girls es obvia y Ginny & Georgia la invita fácilmente, pero ese punto de referencia es simplemente una capa de lo que Ginny & Georgia está haciendo',
 "Llamarlo una imitación de Gilmore sería como tratar de describir los Nachos de bote de basura diciendo: 'Oh, sí, son como la mayoría de los nachos que harías para ti en un día típico",
 "' Porque Ginny & Georgia tiene mucho más que hacer que 

## Librería Evaluate
 

In [ ]:
!pip install evaluate -Uq

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 81.4/81.4 KB 3.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 190.3/190.3 KB 9.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 28.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 213.0/213.0 KB 20.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 132.0/132.0 KB 14.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 110.5/110.5 KB 10.6 MB/s eta 0:00:00


Además necesitamos instalar una serie de librerías adicionales que son necesarias para las métricas que vamos a emplear. 

In [ ]:
!pip install jiwer bert_score git+https://github.com/google-research/bleurt.git -Uq

  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.1/61.1 KB 2.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 27.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 75.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.3/6.3 MB 77.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 352.1/352.1 KB 30.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 58.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.6/7.6 MB 90.7 MB/s eta 0:00:00


### WER




In [ ]:
from evaluate import load
!pip install jiwer
wer = load("wer")

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
wer_score1 = wer.compute(predictions=frasesdeepl, references=frasesReferencia)
wer_score1

0.5207100591715976

In [ ]:
wer_score2 = wer.compute(predictions=frasesyandex, references=frasesReferencia)
wer_score2

0.6035502958579881

### BLEU


In [ ]:
bleu = load("bleu")

In [ ]:
bleu_score1 = bleu.compute(predictions=frasesdeepl, references=frasesReferencia)
bleu_score1


{'bleu': 0.37314597456280785,
 'precisions': [0.6771428571428572,
  0.45722713864306785,
  0.32621951219512196,
  0.21766561514195584],
 'brevity_penalty': 0.9690601725654608,
 'length_ratio': 0.9695290858725761,
 'translation_length': 350,
 'reference_length': 361}

In [ ]:
bleu_score2 = bleu.compute(predictions=frasesyandex, references=frasesReferencia)
bleu_score2

{'bleu': 0.3049595219695165,
 'precisions': [0.6353591160220995,
  0.3732193732193732,
  0.23529411764705882,
  0.15501519756838905],
 'brevity_penalty': 1.0,
 'length_ratio': 1.002770083102493,
 'translation_length': 362,
 'reference_length': 361}

### METEOR


In [ ]:
meteor = load("meteor")

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


In [ ]:
meteor_score1 = meteor.compute(predictions=frasesdeepl, references=frasesReferencia)
meteor_score1

{'meteor': 0.622256547297383}

In [ ]:
meteor_score2 = meteor.compute(predictions=frasesyandex, references=frasesReferencia)
meteor_score2

{'meteor': 0.5832392921334314}

### BERT Score


In [ ]:
!pip install bert_score
bertscore = load("bertscore")

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
bert_score1 = bertscore.compute(predictions=frasesdeepl, references=frasesReferencia,lang="es")
bert_score1

{'precision': [0.9363266229629517,
  0.9065191745758057,
  0.8288123607635498,
  0.7846939563751221,
  0.8602566719055176,
  0.9049326181411743,
  0.8562231659889221,
  0.869369626045227,
  0.9432770013809204,
  0.924179196357727,
  0.8248199820518494],
 'recall': [0.9030449986457825,
  0.8937544226646423,
  0.8332027196884155,
  0.7564681172370911,
  0.8912934064865112,
  0.8707194328308105,
  0.8551961779594421,
  0.8822975754737854,
  0.8979564905166626,
  0.9137584567070007,
  0.8269283175468445],
 'f1': [0.9193846583366394,
  0.9000915288925171,
  0.8310016989707947,
  0.7703226208686829,
  0.8755000829696655,
  0.8874964118003845,
  0.8557093739509583,
  0.8757858276367188,
  0.9200589656829834,
  0.9189393520355225,
  0.8258727788925171],
 'hashcode': 'bert-base-multilingual-cased_L9_no-idf_version=0.3.12(hug_trans=4.26.1)'}

In [ ]:
bert_score2 = bertscore.compute(predictions=frasesyandex, references=frasesReferencia,lang="es")
bert_score2

{'precision': [0.8427435755729675,
  0.9147112369537354,
  0.8554118275642395,
  0.7714084982872009,
  0.8764582872390747,
  0.8531836867332458,
  0.8070610761642456,
  0.861115574836731,
  0.9044003486633301,
  0.9351773858070374,
  0.8073806762695312],
 'recall': [0.8309818506240845,
  0.9032226800918579,
  0.858964741230011,
  0.7383645176887512,
  0.8940605521202087,
  0.8427248001098633,
  0.8316797018051147,
  0.8769657611846924,
  0.8606052398681641,
  0.9412189722061157,
  0.8310236930847168],
 'f1': [0.836821436882019,
  0.9089305996894836,
  0.8571845889091492,
  0.7545248866081238,
  0.8851719498634338,
  0.8479219079017639,
  0.8191854357719421,
  0.8689683675765991,
  0.8819594979286194,
  0.9381884336471558,
  0.8190316557884216],
 'hashcode': 'bert-base-multilingual-cased_L9_no-idf_version=0.3.12(hug_trans=4.26.1)'}

In [ ]:
import numpy as np
np.mean(bert_score1['f1']),np.std(bert_score1['f1'])

(0.8709239363670349, 0.04493798427702927)

In [ ]:
np.mean(bert_score2['f1']),np.std(bert_score2['f1'])

(0.8561717055060647, 0.04738909042505856)

### BLEURT


In [ ]:
!pip install git+https://github.com/google-research/bleurt.git
bleurtscore = load("bleurt")

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Cloning https://github.com/google-research/bleurt.git to /tmp/pip-req-build-5gg2i851
  Running command git clone --filter=blob:none --quiet https://github.com/google-research/bleurt.git /tmp/pip-req-build-5gg2i851
  Resolved https://github.com/google-research/bleurt.git to commit cebe7e6f996b40910cfaa520a63db47807e3bf5c
  Preparing metadata (setup.py) ... done


In [ ]:
bleurt_score1 = bleurtscore.compute(predictions=frasesdeepl, references=frasesReferencia)
bleurt_score1

{'scores': [0.47153663635253906,
  0.06312122941017151,
  -0.039444658905267715,
  -0.20588557422161102,
  0.09242743253707886,
  0.13542518019676208,
  0.11418400704860687,
  0.06512819230556488,
  0.1333075761795044,
  0.3512440323829651,
  0.04864305630326271]}

In [ ]:
bleurt_score2 = bleurtscore.compute(predictions=frasesyandex, references=frasesReferencia)
bleurt_score2

{'scores': [-0.4041782021522522,
  0.14159485697746277,
  0.16056965291500092,
  -0.256451278924942,
  0.07490509748458862,
  -0.02676958218216896,
  0.10717961192131042,
  -0.0795527994632721,
  0.013253871351480484,
  0.3958672881126404,
  0.08945679664611816]}

In [ ]:
np.mean(bleurt_score1['scores']),np.std(bleurt_score1['scores'])

(0.11178973723541606, 0.1706888041927465)

In [ ]:
np.mean(bleurt_score2['scores']),np.std(bleurt_score2['scores'])

(0.019625028425996952, 0.2044404847853853)

#Resumen resultados

|Métrica|DeepL|Yandex|
|---|---|---|
| WER | 0.52 | 0.60|
| BLEU | 0.37 | 0.30|
| METEOR |0.62|0.58|
| BERT Score | 0.87 (0.04) | 0.85 (0.04)|
| BLEURT | 0.11 (0.17) | 0.19 (0.20)|

